File: lintcode/remove-duplicates-from-sorted-list-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/remove-duplicates-from-sorted-list-ii/

from linked_list import *

"""
Definition of ListNode
class ListNode(object):
    def __init__(alf, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: A ListNode
    @return: A ListNode
    """
    def deleteDuplicates(self, head):
        # write your code here
        if not head or not head.next:
            return head

        new_head = None
        writer = None
        reader_slow = head
        reader_fast = head.next
        while reader_fast:
            old_reader_slow = reader_slow
            while reader_fast and reader_slow.val == reader_fast.val:
                reader_slow = reader_fast
                reader_fast = reader_fast.next

            if old_reader_slow == reader_slow:
                if not new_head:
                    new_head = reader_slow
                    writer = new_head
                else:
                    writer.next = reader_slow
                    writer = reader_slow

            if reader_fast:
                reader_slow = reader_fast
                reader_fast = reader_fast.next
                if not reader_fast:
                    writer.next = reader_slow
                    writer = writer.next

            if writer:
                writer.next = None

        return new_head

if __name__ == '__main__':
    s = Solution()
    head = create_linked_list([1,2,3,3,4,4,5])
    new_head = s.deleteDuplicates(head)
    print_linked_list(new_head)
    head = create_linked_list([2,3,4,4,4])
    new_head = s.deleteDuplicates(head)
    print_linked_list(new_head)
    head = create_linked_list([1,1,1])
    new_head = s.deleteDuplicates(head)
    print_linked_list(new_head)

File: leetcode/remove-duplicates-from-sorted-list-ii.cc
// Question: https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

#include "headers.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head) {
            return NULL;
        }

        int num = head->val, cnt = 1;
        ListNode* node = NULL;
        ListNode* cur = head->next;
        while (cur) {
            if (cur->val == num) {
                ++cnt;
            } else {
                if (cnt == 1) {
                    if (!node) {
                        node = head;
                    } else {
                        node = node->next;
                    }
                    node->val = num;
                }
                num = cur->val;
                cnt = 1;
            }
            cur = cur->next;
        }
        if (cnt == 1) {
            if (!node) {
                node = head;
            } else {
                node = node->next;
            }
            node->val = num;
        }
        if (node) {
            node->next = NULL;
            return head;
        } else {
            return NULL;
        }
    }
};

int main(int argc, char** argv) {
    Solution s;
    ListNode* head = new ListNode(1);
    // head->next = new ListNode(2);
    // head->next->next = new ListNode(3);
    // head->next->next->next = new ListNode(3);
    // head->next->next->next->next = new ListNode(4);
    // head->next->next->next->next->next = new ListNode(4);
    // head->next->next->next->next->next->next = new ListNode(5);
    ListNode* new_head = s.deleteDuplicates(head);
    while (new_head) {
        cout << new_head->val << " ";
        new_head = new_head->next;
    }
    cout << endl;
    return 0;
}
